File: src/main/java/com/example/apkdownloader/DownloadManager.java
package com.example.apkdownloader;

import java.util.ArrayList;
import java.util.List;

public class DownloadManager {
    private List<DownloadTask> downloadQueue;
    private List<ProgressListener> progressListeners;

    public DownloadManager() {
        downloadQueue = new ArrayList<>();
        progressListeners = new ArrayList<>();
    }

    public void addDownloadTask(DownloadTask task) {
        downloadQueue.add(task);
        notifyProgressListeners(0);
    }

    public void removeDownloadTask(DownloadTask task) {
        downloadQueue.remove(task);
        notifyProgressListeners(0);
    }

    public void startDownload(DownloadTask task) {
        // Implement download logic here
        // Notify progress listeners during the download process
    }

    public void addProgressListener(ProgressListener listener) {
        progressListeners.add(listener);
    }

    public void removeProgressListener(ProgressListener listener) {
        progressListeners.remove(listener);
    }

    private void notifyProgressListeners(int progress) {
        for (ProgressListener listener : progressListeners) {
            listener.onProgress(progress);
        }
    }

    public List<DownloadTask> getDownloadQueue() {
        return downloadQueue;
    }

    public void setDownloadQueue(List<DownloadTask> downloadQueue) {
        this.downloadQueue = downloadQueue;
    }
}

File: src/main/java/com/example/apkdownloader/DownloadManager.kt
package com.example.apkdownloader

class DownloadManager {
    private val downloadQueue: MutableList<DownloadTask> = mutableListOf()
    private val progressListeners: MutableList<ProgressListener> = mutableListOf()

    fun addDownloadTask(task: DownloadTask) {
        downloadQueue.add(task)
        notifyProgressListeners(0)
    }

    fun removeDownloadTask(task: DownloadTask) {
        downloadQueue.remove(task)
        notifyProgressListeners(0)
    }

    fun startDownload(task: DownloadTask) {
        // Implement download logic here
        // Notify progress listeners during the download process
    }

    fun addProgressListener(listener: ProgressListener) {
        progressListeners.add(listener)
    }

    fun removeProgressListener(listener: ProgressListener) {
        progressListeners.remove(listener)
    }

    private fun notifyProgressListeners(progress: Int) {
        for (listener in progressListeners) {
            listener.onProgress(progress)
        }
    }

    fun getDownloadQueue(): List<DownloadTask> {
        return downloadQueue
    }

    fun setDownloadQueue(downloadQueue: MutableList<DownloadTask>) {
        this.downloadQueue = downloadQueue
    }
}
